Semantics on top of profiles (#115)
WebAssembly/relaxed-simd: document/core/exec/numerics.rst

   \begin{array}{lll@{\qquad}l}
   \narrowu_{M,N}(i) &=& \satu_N(\signed_M(i))
   \end{array}

.. _relaxed-ops:

Relaxed operators
~~~~~~~~~~~~~~~~~

The result of *relaxed* operators are *host-dependent*, because the set of
possible results may depend on properties of the host environment (such as
hardware). Technically, each such operator produces a fixed-size *list of sets*
of allowed values.  For each execution of the operator in the same environment,
only values from the set at the same position in the list are returned, i.e.,
each environment globally chooses a fixed projection for each operator.

.. note::
   Each operator can be thought of as a family of operations that is fixed to
   one particular choice by the execution environment. The fixed operation
   itself can still be non-deterministic or partial.

.. _op-frmadd:

:math:`\frmadd_N(z_1, z_2, z_3)`
................................

Relaxed Multiply Add (madd) allows for fused or unfused results. :math:`fma` is
defined by |IEEE754|_ (Section 5.4.1) as *fusedMultiplyAdd*.

.. math::
   \begin{array}{@{}llcll}
   \EXPROFDET & \frmadd_N(z_1, z_2, z_3) &=& [ \fadd_N(\fmul_N(z_1, z_2), z_3), fma_N(z_1, z_2, z_3) ] \\

>    & \frmadd_N(z_1, z_2, z_3) &=& \fadd_N(\fmul_N(z_1, z_2), z_3) \\

Reviewer: Do I read this correctly as saying that the deterministic semantics are non-fused?

In [the 01-31 meeting](https://github.com/WebAssembly/meetings/blob/main/main/2023/CG-01-31.md), the vote favored that the deterministic semantics should be fused:

ZA: vote for det semantics for relaxed madd/nmadd:
1. (real) fma
2. mul + add (with intermediate rounding)
3. neutral

**POLL**

|FMA|Mul+Add|Neutral|
|---|--|--|
|15|2|16|

Author: This is wrong, needs to be fixed. Thanks for pointing this out.

Author: Fixed in [18c665a](https://github.com/WebAssembly/relaxed-simd/pull/115/commits/18c665adbfce7102ba841c01cd61525dd3b6a7b6)

   \begin{array}{lll@{\qquad}l}
   \narrowu_{M,N}(i) &=& \satu_N(\signed_M(i))
   \end{array}

.. _relaxed-ops:

Relaxed operations
~~~~~~~~~~~~~~~~~~

The result of *relaxed* operators are *host-dependent*, because the set of
possible results may depend on properties of the host environment (such as
hardware). Technically, each such operator produces a fixed-size *list of sets*
of allowed values.  For each execution of the operator in the same environment,
only values from the set at the same position in the list are returned, i.e.,
each environment globally chooses a fixed projection for each operator.

.. note::
   Each operator can be thought of as a family of operations that is fixed to
   one particular choice by the execution environment. The fixed operation
   itself can still be non-deterministic or partial.

.. _op-relaxed_madd:

:math:`\relaxedmadd_N(z_1, z_2, z_3)`
.....................................

> Relaxed multiply-add (madd) allows for fused or unfused results. The function :math:`fma` is

Reviewer 2: ```suggestion
Relaxed multiply-add allows for fused or unfused results. The function :math:`\F{fma}` is
```
(applies to formatting of `fma` here and below, or perhaps a macro)

> defined by |IEEE754|_ (Section 5.4.1) as *fusedMultiplyAdd*.

Reviewer 2: Aren't we still missing a definition for this function?

Author: Done, ptal.

~~~~~~~~~~~~~~~~~~

The result of *relaxed* operators are *host-dependent*, because the set of
possible results may depend on properties of the host environment (such as
hardware). Technically, each such operator produces a fixed-size *list of sets*
of allowed values.  For each execution of the operator in the same environment,
only values from the set at the same position in the list are returned, i.e.,
each environment globally chooses a fixed projection for each operator.

.. note::
   Each operator can be thought of as a family of operations that is fixed to
   one particular choice by the execution environment. The fixed operation
   itself can still be non-deterministic or partial.

.. _op-relaxed_madd:
.. _op-fma:

:math:`\relaxedmadd_N(z_1, z_2, z_3)`
.....................................

Relaxed multiply-add allows for fused or unfused results. The function :math:`\fma`
is the same as *fusedMultiplyAdd* defined by |IEEE754|_ (Section 5.4.1).

.. math::
   \begin{array}{@{}llcll}
   \EXPROFDET & \relaxedmadd_N(z_1, z_2, z_3) &=& [ \fadd_N(\fmul_N(z_1, z_2), z_3), \fma_N(z_1, z_2, z_3) ] \\
   & \relaxedmadd_N(z_1, z_2, z_3) &=& \fma_N(z_1, z_2, z_3) \\
   \\

>    & \fma_N(z_1, z_2, z_3) &=& \ieee_N(z_1 \cdot z_2 + z_3) \\

Reviewer 2: Ah, sorry, I was a bit quick to approve. This of course needs to specify all the cases for NaNs and infinities, in the same manner as fadd etc.

(Also, since I didn't point it out explicitly before, please don't forget that all the descriptions here still need a proper prose counterpart.)

Author: Yea, I missed it too, ptal #135.

   one particular choice by the execution environment. The fixed operation
   itself can still be non-deterministic or partial.

.. _op-frmadd:

:math:`\frmadd_N(z_1, z_2, z_3)`
................................

> Relaxed Multiply Add (madd) allows for fused or unfused results. :math:`fma` is
> defined by |IEEE754|_ (Section 5.4.1) as *fusedMultiplyAdd*.

Reviewer 2: ```suggestion
Relaxed multiply-add allows for fused or unfused results. The function :math:`\X{fma}` is
defined by |IEEE754|_ (Section 5.4.1) as *fusedMultiplyAdd*.
```
Despite this citation, I'm afraid we need a full definition here. The Wasm spec is ought to be self-contained.

Author: Hm, IEEE754 doesn't really specify this, just says "computed to unbounded precision"
@Maratyszcza any ideas how we can specify FMA here?

Author: https://webassembly.github.io/spec/core/exec/numerics.html#floating-point-operations
The spec defines the floating point format, but not floating point arithmetics operations, can we similarly punt fusedMultiplyAdd definition to IEEE-754?

Reviewer 2: The spec specifies all arithmetic operations to the last bit, mainly in terms of real arithmetics and rounding. This should suffice for FMA as well, if I understand its semantics correctly. (The more involved part probably is writing suitable tests.)

Author: I have some tests with special values to check for FMA https://github.com/WebAssembly/relaxed-simd/blob/main/test/core/relaxed-simd/relaxed_madd_nmadd.wast

.. math::
   \begin{array}{@{}llcll}
   \EXPROFDET & \frmadd_N(z_1, z_2, z_3) &=& [ \fadd_N(\fmul_N(z_1, z_2), z_3), fma_N(z_1, z_2, z_3) ] \\
   & \frmadd_N(z_1, z_2, z_3) &=& \fadd_N(\fmul_N(z_1, z_2), z_3) \\
   \end{array}

.. _op-frnmadd:

:math:`\frnmadd_N(z_1, z_2, z_3)`
.................................

> Relaxed Negative Multiply Add (nmadd) allows for fused or unfused results.

Reviewer 2: ```suggestion
Relaxed negative multiply-add allows for fused or unfused results.
```

.. math::
   \begin{array}{@{}llcll}
   & \frnmadd_N(z_1, z_2, z_3) &=& \frmadd_N(-z_1, z_2, z_3) \\
   \end{array}

.. _op-rswizzle:

:math:`\rswizzle(a^n, s^n)`
...........................

Relaxed swizzle lane is a helper for relaxed swizzle. Result is deterministic
if the signed interpretation of the index is less than 16 (including negative).
:math:`j` is a 8-bit int.

.. math::
   \begin{array}{@{}llcll}
   & relaxed\_swizzle\_lane(i^n, j) &=& i[j] & (\iff j < 16) \\
   & relaxed\_swizzle\_lane(i^n, j) &=& 0 & (\iff \signed_8(j) < 0) \\
   \EXPROFDET & relaxed\_swizzle\_lane(i^n, j) &=& [ 0, i[j \mod n] ] & (\otherwise) \\
   & relaxed\_swizzle\_lane(i^n, j) &=& 0 & (\otherwise) \\
   \\
   & \rswizzle(a^n, s^n) &=& rsl_0 \dots rsl_{n-1} \\
   & \qquad \where rsl_i &=& relaxed\_swizzle\_lane(a^n, s^n[i])
   \end{array}

.. _op-rtrunc:
.. _op-rtrunc_u:

:math:`\rtrunc^u_{M,N}(z)`
..........................

Relaxed truncate converts float to int, NaN and out of range values are

> hardware dependent.

Reviewer 2: ```suggestion
Relaxed unsigned truncation converts floating point values to integers. The result for NaN's and out-of-range values is implementation-dependent.
```

.. math::
   \begin{array}{@{}llcll}
   \EXPROFDET & \rtrunc^u_{M,N}(\pm \NAN(n)) &=& [ 0, 2^{N}-1 ] \\
   \EXPROFDET & \rtrunc^u_{M,N}(\pm q) &=& \truncu_{M,N}(\pm q) & (\iff -1 < \trunc(\pm q) < 2^N) \\
   \EXPROFDET & \rtrunc^u_{M,N}(\pm p) &=& [ \truncsatu_{M,N}(\pm p), 2^{N}-1 ] & (\otherwise) \\
   & \rtrunc^u_{M,N}(\pm p) &=& \truncsatu_{M,N}(\pm p) & (\otherwise) \\
   \end{array}

.. _op-rtrunc_s:

:math:`\rtrunc^s_{M,N}(z)`
..........................

> Relaxed truncate converts float to int, NaN and out of range values are
> hardware dependent.

Reviewer 2: ```suggestion
Relaxed signed truncation converts floating point numbers to integers.
The result for NaN's and out-of-range values is host-dependent.
```

.. math::
   \begin{array}{@{}llcll}
   \EXPROFDET & \rtrunc^s_{M,N}(\pm \NAN(n)) &=& [ 0, -2^{N-1} ] \\
   \EXPROFDET & \rtrunc^s_{M,N}(\pm q) &=& \truncs_{M,N}(\pm q) & (\iff -2^{N-1} - 1 < \trunc(\pm q) < 2^{N-1}) \\
   \EXPROFDET & \rtrunc^s_{M,N}(\pm p) &=& [ \truncsats_{M,N}(\pm p), -2^{N-1} ] & (\otherwise) \\
   & \rtrunc^s_{M,N}(\pm p) &=& \truncsats_{M,N}(\pm p) & (\otherwise) \\
   \end{array}

.. _op-rlaneselect:

:math:`\rlaneselect_B(a^n, b^n, c^n)`
.....................................

Relaxed lane select is deterministic where all bits are set or unset in the
mask. Otherwise depending on hardware, either only the top bit is examined, or

> all bits are examined (becomes a bitselect).

Reviewer 2: ```suggestion
Relaxed lane selection is deterministic when all bits are set or unset in the
mask. Otherwise depending on the host, either only the top bit is examined, or
all bits are examined (i.e., it becomes a bit select).
```

.. math::
   \begin{array}{@{}llcll}

>    & relaxed\_lane_N(i_1, i_2, 2^N-1) &=& i_1 \\

Reviewer 2: Use a macro.

   & relaxed\_lane_N(i_1, i_2, 0) &=& i_2 \\
   \EXPROFDET & relaxed\_lane_N(i_1, i_2, i_3) &=& [ \ibitselect_N(i_1, i_2, i_3), i_1 ] & (\iff \signed_N(i_3) < 0) \\
   \EXPROFDET & relaxed\_lane_N(i_1, i_2, i_3) &=& [ \ibitselect_N(i_1, i_2, i_3), i_2 ] & (\otherwise) \\
   & relaxed\_lane_N(i_1, i_2, i_3) &=& \ibitselect_N(i_1, i_2, i_3) & (\otherwise) \\
   \\
   & \rlaneselect_B(a^n, b^n, c^n) &=& rll_0 \dots rll_{n-1} \\
   & \qquad \where rll_i &=& relaxed\_lane_B(a^n[i], b^n[i], c^n[i]) \\
   \end{array}

.. _op-frmin:

:math:`\frmin_N(z_1, z_2)`
..........................

Relaxed min differs from min when inputs are NaNs or different
signs of 0. It allows for implementation to return the first or second input

> when either input is a NaN.

Reviewer 2: ```suggestion
Relaxed minimum differs from regular minimum when inputs are NaN's or zeroes with different signs.
It allows for implementations to return the first or second input when either input is a NaN.
```

.. math::
   \begin{array}{@{}llcll}
   \EXPROFDET & \frmin_N(\pm \NAN(n), z_2) &=& [ \nans_N\{\pm \NAN(n), z_2\}, \NAN(n), z_2, z_2 ] \\
   \EXPROFDET & \frmin_N(z_1, \pm \NAN(n)) &=& [ \nans_N\{\pm \NAN(n), z_1\}, z_1, \NAN(n), z_1 ] \\
   \EXPROFDET & \frmin_N(\pm 0, \mp 0) &=& [ -0, \pm 0, \mp 0, -0 ] \\
   & \frmin_N(z_1, z_2) &=& \fmin_N(z_1, z_2) & (\otherwise) \\
   \end{array}

.. _op-frmax:

:math:`\frmax_N(z_1, z_2)`
..........................

> Relaxed max differs from max when inputs are NaNs or different
> signs of 0. It allows for implementation to return the first or second input
> when either input is a NaN.

Reviewer 2: ```suggestion
Relaxed maximum differs from regular maximum when inputs are NaN's or zeroes with different signs.
It allows for implementations to return the first or second input when either input is a NaN.
```

.. math::
   \begin{array}{@{}llcll}
   \EXPROFDET & \frmax_N(\pm \NAN(n), z_2) &=& [ \nans_N\{\pm \NAN(n), z_2\}, \NAN(n), z_2, z_2 ] \\
   \EXPROFDET & \frmax_N(z_1, \pm \NAN(n)) &=& [ \nans_N\{\pm \NAN(n), z_1\}, z_1, \NAN(n), z_1 ] \\
   \EXPROFDET & \frmax_N(\pm 0, \mp 0) &=& [ +0, \pm 0, \mp 0, +0 ] \\
   & \frmax_N(z_1, z_2) &=& \fmax_N(z_1, z_2) & (\otherwise) \\
   \end{array}

.. _op-irq15mulr_s:

:math:`\irq15mulrs_N(i_1, i_2)`
...............................

.. math::
   \begin{array}{@{}llcll}
   \EXPROFDET & \irq15mulrs_N(i_1, i_2) &=& [ \sats_N(i), i \mod 2^N ] & (\iff i = \ishrs_N(i_1 \cdot i_2 + 2^{14}, 15)) \\
   & \irq15mulrs_N(i_1, i_2) &=& \iq15mulrsats(i_1, i_2)
   \end{array}

.. _op-ridotmul:

:math:`\ridotmul_{M,N}(i_1, i_2)`
.................................

> Relaxed integer dot product differs from integer dot product when the elements of the input have top bit set.

Reviewer 2: ```suggestion
Relaxed integer dot product differs from regular integer dot product when the elements of the input have their most significant bit set.
```
